File: apps/somi-home-android/app/src/main/java/com/somi/home/core/models/ApiModels.kt
package com.somi.home.core.models

import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class LoginResponse(
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Int,
    val user: UserInfo
)

@JsonClass(generateAdapter = true)
data class UserInfo(
    val userId: String,
    val email: String,
    val role: String,
    val patientId: String?,
    val displayName: String?
)

@JsonClass(generateAdapter = true)
data class MeResponse(
    val userId: String,
    val email: String,
    val role: String,
    val patientId: String?,
    val displayName: String?
)

@JsonClass(generateAdapter = true)
data class RefreshRequest(val refreshToken: String)

@JsonClass(generateAdapter = true)
data class TokenResponse(
    val accessToken: String,
    val refreshToken: String,
    val expiresIn: Int
)

@JsonClass(generateAdapter = true)
data class TodayViewResponse(
    val dateLocal: String,
    val sessionKey: String,
    val sessionTitle: String?,
    val sessionNotes: String?,
    val timesPerDay: Int,
    val assignments: List<TodayAssignment>,
    val overallCompletionRate: Double
)

@JsonClass(generateAdapter = true)
data class TodayAssignment(
    val assignmentKey: String,
    val exerciseVersionId: String,
    val exercise: ExerciseInfo,
    // Backend returns pre-merged params under this key (default + any plan override applied).
    val effectiveParams: ExerciseParams?,
    val completions: List<CompletionEntry>
)

@JsonClass(generateAdapter = true)
data class ExerciseInfo(
    val title: String,
    val description: String,
    // Absent from today-view responses (backend pre-merges into effectiveParams);
    // present in plan-view responses.
    val defaultParams: ExerciseParams?,
    val mediaId: String?
)

@JsonClass(generateAdapter = true)
data class ExerciseParams(
    val reps: Int?,
    val sets: Int?,
    val seconds: Int?
)

@JsonClass(generateAdapter = true)
data class CompletionEntry(
    val occurrence: Int,
    val completed: Boolean,
    val completedAt: String? // null when not yet completed
)

@JsonClass(generateAdapter = true)
data class TreatmentPlan(
    val planId: String,
    val patientId: String,
    val status: String,
    val sessions: List<PlanSession>
)

@JsonClass(generateAdapter = true)
data class PlanSession(
    val sessionKey: String,
    val index: Int,
    val title: String?,
    val timesPerDay: Int,
    val assignments: List<PlanAssignment>
)

@JsonClass(generateAdapter = true)
data class PlanAssignment(
    val assignmentKey: String,
    val exerciseVersionId: String,
    val index: Int,
    val exercise: ExerciseInfo?,
    val paramsOverride: ExerciseParams?
)

@JsonClass(generateAdapter = true)
data class MessageThread(
    val threadId: String,
    val patientId: String,
    val therapistUserId: String,
    val status: String,
    val lastMessageAt: String?
)

@JsonClass(generateAdapter = true)
data class Message(
    val messageId: String,
    val threadId: String,
    val senderUserId: String,
    val senderRole: String,
    val text: String,
    val createdAt: String
)

@JsonClass(generateAdapter = true)
data class PagedMessages(
    val items: List<Message>,
    val nextCursor: String?
)

@JsonClass(generateAdapter = true)
data class SendMessageRequest(
    val text: String,
    val attachmentUploadIds: List<String> = emptyList()
)

@JsonClass(generateAdapter = true)
data class CompletionRequest(
    val dateLocal: String,
    val occurrence: Int,
    val exerciseVersionId: String,
    val source: String
)

@JsonClass(generateAdapter = true)
data class UncompletionRequest(
    val dateLocal: String,
    val occurrence: Int,
    val exerciseVersionId: String
)

@JsonClass(generateAdapter = true)
data class CompletionResponse(
    val completionId: String,
    val completedAt: String
)

@JsonClass(generateAdapter = true)
data class VideoAccessResponse(
    val accessUrl: String,
    val expiresAt: String
)

File: apps/somi-home-ios/SOMIHome/Core/Models/APIModels.swift
import Foundation

// MARK: - Auth

struct LoginResponse: Codable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
    let user: UserInfo
}

struct UserInfo: Codable {
    let userId: String
    let email: String
    let role: String
    let patientId: String?
    let displayName: String?
}

struct MeResponse: Codable {
    let userId: String
    let email: String
    let role: String
    let patientId: String?
    let displayName: String?
}

struct RefreshResponse: Codable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
}

// MARK: - Today View
// Flat single-session response from GET /v1/me/today.
// effectiveParams are merged server-side; completions pre-populated for each occurrence.

struct TodayViewResponse: Codable {
    let dateLocal: String
    let sessionKey: String
    let sessionTitle: String?
    let sessionNotes: String?
    let timesPerDay: Int
    var assignments: [TodayAssignment]
    let overallCompletionRate: Double
}

struct TodayAssignment: Codable, Identifiable {
    var id: String { assignmentKey }
    let assignmentKey: String
    let exerciseVersionId: String
    let exercise: ExerciseInfo?
    let effectiveParams: ExerciseParams
    var completions: [CompletionEntry]
}

struct ExerciseInfo: Codable {
    let title: String
    let description: String
    let mediaId: String?
    let defaultParams: ExerciseParams?  // Present in Plan API; nil in Today API (params are pre-merged into effectiveParams)
}

struct ExerciseParams: Codable {
    let reps: Int?
    let sets: Int?
    let seconds: Int?
}

struct CompletionEntry: Codable {
    let occurrence: Int
    let completed: Bool
    let completedAt: String?
}

// MARK: - Treatment Plan

struct TreatmentPlan: Codable {
    let planId: String
    let patientId: String
    let status: String
    let sessions: [PlanSession]
}

struct PlanSession: Codable, Identifiable {
    var id: String { sessionKey }
    let sessionKey: String
    let index: Int
    let title: String?
    let sessionNotes: String?
    let timesPerDay: Int
    let assignments: [PlanAssignment]
}

struct PlanAssignment: Codable, Identifiable {
    var id: String { assignmentKey }
    let assignmentKey: String
    let exerciseVersionId: String
    let index: Int
    let exercise: ExerciseInfo?
    let paramsOverride: ExerciseParams?
}

// MARK: - Messages

struct MessageThread: Codable {
    let threadId: String
    let patientId: String
    let therapistUserId: String
    let status: String
    let lastMessageAt: String?
}

struct Message: Codable, Identifiable {
    var id: String { messageId }
    let messageId: String
    let threadId: String
    let senderUserId: String
    let senderRole: String
    let text: String
    let createdAt: String
}

struct PagedMessages: Codable {
    let items: [Message]
    let nextCursor: String?
}

struct SendMessageRequest: Codable {
    let text: String
    let attachmentUploadIds: [String]
}

// MARK: - Uploads

struct VideoAccessResponse: Codable {
    let accessUrl: String
    let expiresAt: String
}

// MARK: - Completions

struct CompletionRequest: Codable {
    let dateLocal: String
    let occurrence: Int
    let exerciseVersionId: String
    let source: String
}

struct CompletionResponse: Codable {
    let completionId: String
    let completedAt: String
}
